components/servo/generic/servo.go

```diff
@@ -0,0 +1,337 @@
+	}
+	pin, err := b.GPIOPinByName(attr.Pin)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get servo pin")
+	}
+	frequency, err := pin.PWMFreq(ctx, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get servo pin pwm frequency")
+	}
+	if attr.Frequency != nil {
+		err = pin.SetPWMFreq(ctx, *attr.Frequency, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "error setting servo pin frequency")
+		}
+		frequency = *attr.Frequency
+	}
+	if frequency > 500 {
+		return nil, errors.Errorf("PWM frequencies should not be above 500Hz, have %d", frequency)
+	}
+	min := 0.0
+	max := 180.0
+	if attr.MinDeg != nil {
+		min = *attr.MinDeg
+	}
+	if attr.MaxDeg != nil {
+		max = *attr.MaxDeg
+	}
+	startPos := 0.0
+	if attr.StartPos != nil {
+		startPos = *attr.StartPos
+	}
+	minUs := minWidthUs
+	maxUs := maxWidthUs
+	if attr.MinWidthUS != nil {
+		minUs = *attr.MinWidthUS
+	}
+	if attr.MaxWidthUS != nil {
+		maxUs = *attr.MaxWidthUS
+	}
+	servo := &servoGeneric{
+		min:       min,
+		max:       max,
+		frequency: frequency,
+		pin:       pin,
+		logger:    logger,
+		minUs:     minUs,
+		maxUs:     maxUs,
+		currPct:   0,
+	}
+	if err := servo.Move(ctx, uint8(startPos), nil); err != nil {
+		return nil, errors.Wrap(err, "couldn't move servo to start position")
+	}
+	if servo.pwmRes == 0 {
+		if err := servo.findPWMResolution(ctx); err != nil {
+			return nil, errors.Wrap(err, "failed to guess the pwm resolution")
+		}
+		if err := servo.Move(ctx, uint8(startPos), nil); err != nil {
+			return nil, errors.Wrap(err, "couldn't move servo to start position")
+		}
+	}
+	return servo, nil
+}
+
+var _ = servo.LocalServo(&servoGeneric{})
+
+// Given minUs,maxUs,deg and frequency attempt to calculate the corresponding duty cylce pct.
+func mapDegToDutyCylePct(minUs, maxUs uint, deg float64, frequency uint) float64 {
+	period := 1.0 / float64(frequency) // dutyCycle in s
+	lSpan := maxDeg - minDeg           // servo moves from 0 to 180 deg
+	rSpan := float64(maxUs - minUs)    // pulse width between 1ms to 2ms to s
+
+	scale := rSpan / lSpan
+
+	pwmWidthUs := float64(minUs) + (deg-minDeg)*scale
+	return (pwmWidthUs / (1000 * 1000)) / period
+}
+
+// Given minUs,maxUs,deg and frequency returns the corresponding duty cycle pct.
+func mapDutyCylePctToDeg(minUs, maxUs uint, pct float64, frequency uint) float64 {
+	period := 1.0 / float64(frequency) // dutyCycle in s
+	pwmWidthus := pct * period * 1000 * 1000
+	rSpan := maxDeg - minDeg        // servo moves from 0 to 180 deg
+	lSpan := float64(maxUs - minUs) // pulse width between 1ms to 2ms to s
+
+	if pwmWidthus < float64(minUs) {
+		pwmWidthus = float64(minUs)
+	}
+	if pwmWidthus > float64(maxUs) {
+		pwmWidthus = float64(maxUs)
+	}
+
+	scale := rSpan / lSpan
+
+	angle := math.Round(minDeg + (pwmWidthus-float64(minUs))*scale)
+
+	return angle
+}
+
+// Attempt to find the PWM resolution assuming an hardware PWM
+// Assumption :
+//  1. 16,15,14,12,or 8 bit timer
```

Comment:
I have no idea what this means. Was this line supposed to be removed before the PR went out? If not, put in more words to explain it better.